File: 461.hamming-distance.cpp
/*
 * @lc app=leetcode id=461 lang=cpp
 *
 * [461] Hamming Distance
 */

#include <bits/stdc++.h>
#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  int hammingDistance(int x, int y) {
    int ret = 0;
    for (int i = 0; i < 32; i++) {
      if ((x^y)&(1<<i)) {
        ret++;
      }
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  assert(s.hammingDistance(1, 4) == 2);
  assert(s.hammingDistance(0, 0) == 0);
  assert(s.hammingDistance(4, 1) == 2);
  assert(s.hammingDistance(4, 4) == 0);
  assert(s.hammingDistance(93, 73) == 2);
  assert(s.hammingDistance(1, 0x40000000) == 2);
  return 0;
}

File: 461.hamming-distance.rs
/*
 * @lc app=leetcode id=461 lang=rust
 *
 * [461] Hamming Distance
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn hamming_distance(x: i32, y: i32) -> i32 {
        let mut ret = 0;
        for i in 0..32 {
            if (x ^ y) & (1 << i) != 0 {
                ret += 1;
            }
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::hamming_distance(1, 4), 2);
}

#[test]
fn test2() {
    assert_eq!(Solution::hamming_distance(3, 1), 1);
}
